sale_order_lot_selection/model/sale.py

```diff
@@ -45,24 +55,28 @@ def product_id_change_with_wh(
 
         available_lots = []
         lot_model = self.env['stock.production.lot']
-        product_model = self.env['product.product']
-        product_obj = product_model.browse(product)
-        location = self.env['stock.warehouse'].browse(
-            warehouse_id).lot_stock_id
-        lots = lot_model.search(
-            [('product_id', '=', product)])
+        product = self.env['product.product'].browse(product_id)
+        warehouse = self.env['stock.warehouse'].browse(warehouse_id)
+        location = self._select_lot_stock_location(product, warehouse)
+        # DO NOT use 'product_id' AFTER THIS LIMIT ; only use 'product'
+        product = self._select_lot_product(product)
+
+        lots = lot_model.search([('product_id', '=', product.id)])
         for lot in lots:
             # for the selected product, search for every associated lot
             # for every lot, check if it is available (in location.id)
             # if it is, add it to selectable lots
-            qty_res = product_obj.with_context({
+            qty_res = product.with_context({
                 'lot_id': lot.id,
                 'location': location.id,
                 })._product_available()
-            if qty_res[product]['qty_available'] > 0:
+            if qty_res[product.id]['qty_available'] > 0:
                 if lot.id not in available_lots:
                     available_lots.append(lot.id)
-        res.update({'domain': {'lot_id': [('id', 'in', available_lots)]}})
+        res.update({'domain': {'lot_id': [
+            ('id', 'in', available_lots),
+            ('product_id', '=', product.id),
+            ]}})
```

Comment:
I tested with this piece of code and seems to works :ok_hand: 
Please look at the setdefault part in order to maintain previous values
```
    @api.multi
    def product_id_change_with_wh(
            self, pricelist, product_id, qty=0,
            uom=False, qty_uos=0, uos=False, name='', partner_id=False,
            lang=False, update_tax=True, date_order=False, packaging=False,
            fiscal_position=False, flag=False, warehouse_id=False):
        res = super(SaleOrderLine, self).product_id_change_with_wh(
            pricelist, product_id, qty=qty, uom=uom,
            qty_uos=qty_uos, uos=uos, name=name, partner_id=partner_id,
            lang=lang, update_tax=update_tax,
            date_order=date_order, packaging=packaging,
            fiscal_position=fiscal_position, flag=flag,
            warehouse_id=warehouse_id)

        product = self.env['product.product'].browse(product_id)
        warehouse = self.env['stock.warehouse'].browse(warehouse_id)
        location = self._select_lot_stock_location(product, warehouse)
        # DO NOT use 'product_id' AFTER THIS LIMIT ; only use 'product'
        product = self._select_lot_product(product)

        # Search all lot existing lot for the product and location selected
        quants = self.env['stock.quant'].read_group([
            ('product_id', '=', product.id),
            ('location_id', 'child_of', location.id),
            ('qty', '>', 0),
            ('lot_id', '!=', False),
            ], ['lot_id'], 'lot_id')
        available_lots = [quant['lot_id'][0] for quant in quants]
        res.setdefault('domain', {}).update(
            {'lot_id': [('id', 'in', available_lots)]})
        res['value']['lot_id'] = False
        return res
```